---
title: "Figure 2 enrichment"
author: "Health Data Science Unit"
date: "`r date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: false
params:
  
  SE: "analysis/tumor/SE_annot/tumor_consensusSE_target_GRanges.RDS"
  terms: "results/supplables/GO_BP_enrichment_SE_target_genes.txt"
  wChIP_tumor : "analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS"
  wChIP_cells : "analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_Wmatrix_Wnorm.RDS"
  wRNAs_tumor : "analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS"
  wRNAv_tumor : "results/figure2/figure2panels_signature_enrichment.pdf"
  figure : "results/figure2/figure2panels_signature_enrichment.pdf"
  sup_figure : "results/sup_figure2/sup_figure2_MostVariable_signature_enrichment.pdf"
  targetannot_df : "analysis/tumor/SE_annot/tumor_consensusSE_target_annotation_df.RDS"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
#this chunk sets global options for Rmarkdown
knitr::opts_chunk$set(echo = TRUE)
#setting working directory. Should be home directory of the project
options(knitr.table.format="html")

library(tidyverse)
library(viridis)
library(ggplot2)
library(patchwork)
library(ggbeeswarm)
library(matrixStats)


```



## Data loading
  
  
```{r dataloading}

# params <- list(SE = "analysis/tumor/SE_annot/tumor_consensusSE_target_GRanges.RDS",
#                terms = "results/supptables/GO_BP_enrichment_SE_target_genes.txt",
#                wChIP_cells = "analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_Wmatrix_Wnorm.RDS",
#                wChIP_tumor = "analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS",
#                wRNAs_tumor = "analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS",
#                wRNAv_tumor = "results/figure2/figure2panels_signature_enrichment.pdf",
#                figure = "results/figure2/figure2panels_signature_enrichment.pdf",
#                sup_figure = "results/sup_figure2/sup_figure2_MostVariable_signature_enrichment.pdf")

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                                 GO gene sets                               ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

# Read table of enriched terms and retrieve unique gene IDs
enriched_terms <- read.table(params$terms, header = TRUE, sep = "\t", stringsAsFactors = FALSE)
#enriched_terms <- enriched_terms[enriched_terms$P.value < 0.01,]
gene_sets <- lapply(split(enriched_terms, enriched_terms$Term), function(termClass){
  gene_list <- termClass$Genes
  unique(unlist(strsplit(gene_list, split = ";")))
})
head(gene_sets)

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                             Read data                                      ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
SE <- readRDS(params$SE)
targetannot_df <- readRDS(params$targetannot_df)

wChIP_cells <- readRDS(params$wChIP_cells)
wChIP_tumor <- readRDS(params$wChIP_tumor)
wRNAs_tumor <- readRDS(params$wRNAs_tumor)
wRNAv_tumor <- readRDS(params$wRNAv_tumor)






```


```{r}

fisher.test.Signatures.Sets <- function(wmatrix, SE, sets, gene.symbols, Specific=TRUE, view="ChIPseq"){
  # Top features
  nmf.feat <- lapply(setNames(colnames(wmatrix), colnames(wmatrix)), function(sigID){
    selec_wmatrix <- do.call(cbind, lapply(as.data.frame(wmatrix), function(sign_expo){
      sign_expo[sign_expo < quantile(sign_expo, 0.90)] <- NA
      sign_expo
    }))
    rownames(selec_wmatrix) <- rownames(wmatrix)
    selec_wmatrix <- selec_wmatrix[!is.na(selec_wmatrix[,sigID]),,drop=FALSE]
    # Keep only the top feature if there's an overlap
    sig_SE_IDs <- rownames(selec_wmatrix[rowMaxs(selec_wmatrix, na.rm = TRUE) == selec_wmatrix[,sigID],])
    
    
    # sign_expo  <- wmatrix[,sigID]
    # sig_SE_IDs <- names(sign_expo[sign_expo > quantile(sign_expo, 0.80)])
    
    if(view=="RNAseq"){
      unique(targetannot_df$SYMBOL[match(sig_SE_IDs, targetannot_df$ENSEMBL)])
      #unique(SE$target_SYMBOL[match(sig_SE_IDs, SE$target_ENSEMBL)])
    } else if (view=="ChIPseq"){
      unique(SE$target_SYMBOL[match(sig_SE_IDs, SE$ID)])
    }
  })
  
  print(lapply(nmf.feat, length))
  #Fisher test
  fish.res <- lapply(nmf.feat, function(signa.feat){
    sapply(sets, function(test.set){
      #Genes in Signature and in the set
      SigYes.SetYes <- sum(signa.feat %in% test.set)
      #Genes in the Signatures Not present in the set
      SigYes.SetNo <- sum(!signa.feat %in% test.set)
      #Genes that are not in the Signture
      SigNo <- gene.symbols[!gene.symbols %in% signa.feat]
      #Genes not present in Signature, but present in the set
      SigNo.SetYes <- sum(SigNo %in% test.set)
      #Genes not present in Signature and set
      SigNo.SetNo <- sum(!SigNo %in% test.set)
      
      test.matrix <- matrix(c(SigYes.SetYes,
                              SigYes.SetNo,
                              SigNo.SetYes,
                              SigNo.SetNo),
                            nrow=2,
                            dimnames=list(Set=c("anno", "NotAnno"),
                                          SignatureFeature=c("Yes", "No")))
      #print(test.matrix)
      fisher.test(test.matrix, alternative = "greater")$p.val
    })
  })
  #return(fish.res)
  fish.res <- do.call(cbind, fish.res)
  fish.df <- as.data.frame(-log10(fish.res))
  fish.df$Set <- rownames(fish.df)
  fish.df <- gather(fish.df, Signature, p.val, -Set)
  fish.df$Signature <- factor(fish.df$Signature, levels = rev(unique(fish.df$Signature)))
  # Map Class of the terms
  fish.df$Class <- enriched_terms$Class[match(fish.df$Set, enriched_terms$Term)]
  #return(fish.df)
  #print(head(fish.df))
  
  
  equal_breaks <- function(n = 3, s = 0.05, ...){
    function(x){
      # rescaling
      d <- s * diff(range(x)) / (1+2*s)
      #round(seq(min(x)+d, max(x)-d, length=n))
      # start axis on 1
      c(1, round(seq(min(x)+d, max(x)-d, length=n))[2])
    }
  }
  
  #print(fish.df)
  
  if (ncol(wmatrix) == 4) {
    colorscale <- c("firebrick", "#00afaf", "#420A68FF", "#2FB47CFF")
  } else {
    colorscale <- c("firebrick", "#420A68FF", "#2FB47CFF")
  }
  
  ggplot(fish.df, aes(x = Signature, y = p.val, color = Signature))+
    geom_boxplot(outlier.size=NA, outlier.shape = NA) +
    geom_quasirandom(size = .5)+
    scale_y_continuous(breaks = equal_breaks(n=2, s=0.05), limits = c(0, NA)) +
    
    facet_grid(.~Class, scales = "free") +
    geom_hline(yintercept = 1.3) +
    ylab(bquote("-log"[10] ~ "(p-value)") ) +
    #scale_color_manual(values = c("#2FB47CFF", "#2FB47CFF", "#420A68FF", "#2FB47CFF")) +
    scale_color_manual(values = colorscale) +
    #scale_color_hue(direction = -1, h.start=90) +
    #scale_color_brewer(type = 'qual', palette = 'Dark2') +
    coord_flip()+
    theme_bw() + 
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), 
          panel.border = element_rect(colour="black"), 
          legend.position = "none"
    )
  
  
  
}



```




```{r}
ggChIP_tumor <- fisher.test.Signatures.Sets(wChIP_tumor, SE, gene_sets, unique(SE$target_ENSEMBL), Specific=T)
ggChIP_cells <- fisher.test.Signatures.Sets(wChIP_cells, SE, gene_sets, unique(SE$target_ENSEMBL), Specific=T)
ggRNAs_tumor <- fisher.test.Signatures.Sets(wRNAs_tumor, SE, gene_sets, unique(SE$target_ENSEMBL), Specific=T, view="RNAseq")
ggRNAv_tumor <- fisher.test.Signatures.Sets(wRNAv_tumor, SE, gene_sets, unique(SE$target_ENSEMBL), Specific=T, view="RNAseq")



main <- ggChIP_tumor + ggChIP_cells+ ggRNAs_tumor + plot_layout(ncol = 1)
sup <- ggRNAv_tumor

main
sup

ggsave(params$figure, main, width = 3, height = 6)
ggsave(params$sup_figure, sup, width = 6, height = 5)


# ggsave(params$figure, main, width = 3, height = 15)
# ggsave(params$sup_figure, sup, width = 4, height = 5)


```

