---
title: "Figure 2h - expression mapped to Mouse GSE99933 E12.5"
author: "Health Data Science Unit"
date: "`r date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: false
params:
  SE_target    : "analysis/tumor/SE_annot/tumor_consensusSE_target_GRanges.RDS"
  wc_tumor     : "analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS"
  wr_tumor     : "analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS"
  mouse_pstime : "db/GSE99933_E12.5/GSE99933_E12.5.txt"
  mouse_exprs  : "db/GSE99933_E12.5/GSE99933_E12.5_exprs_Zscore.txt"
  figure : "results/figure2/figure2h_SEtargetExprs_mouseE12.5.pdf"

editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
#this chunk sets global options for Rmarkdown
knitr::opts_chunk$set(echo = TRUE)
#setting working directory. Should be home directory of the project
options(knitr.table.format="html")

library(tidyverse)
library(viridis)
library(ggplot2)
library(biomaRt)
library(scales)
library(patchwork)
library(rtracklayer)
library(matrixStats)

```



## Data loading
  
  
```{r dataloading}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                             Read data                                      ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# params = list(SE_target = "analysis/tumor/SE_annot/tumor_consensusSE_target_GRanges.RDS",
#               w_tumor = 'analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS',
#               f_tumor      = 'analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_NMF_features.RDS',
#               mouse_pstime = 'db/GSE99933_E12.5/GSE99933_E12.5.txt',
#               mouse_exprs  = 'db/GSE99933_E12.5/GSE99933_E12.5_exprs_Zscore.txt',
#               figure = 'results/figure2/figure2h_SEtargetExprs_mouseE12.5.pdf')

# Read SE annotation
SE <- readRDS(params$SE_target)

# Read NMF results
wc_tumor <- readRDS(params$wc_tumor)
wr_tumor <- readRDS(params$wr_tumor)



# Read Mouse GSE99933 E12.5 data
# Pseudotime and tsne axis
e12.5_pstime <- read.table(params$mouse_pstime, 
                           header = TRUE,
                           stringsAsFactors = FALSE)

head(e12.5_pstime)
# Zscore of gene expression
e12.5_exprs <- read.table(params$mouse_exprs, 
                          header = TRUE,
                          colClasses = c(probeset = "NULL"),
                          row.names = 1,
                          stringsAsFactors = FALSE)
e12.5_exprs[1:5,1:5]


mouseE12.5_NMF_umap <- readRDS("db/GSE99933_E12.5/UMAP_projections/mouseE12.5_NMF_umap.RDS")

e12.5_pstime <- e12.5_pstime[!is.na(e12.5_pstime$tsne_v1),]

match(e12.5_pstime$SampleID, mouseE12.5_NMF_umap$annot$CellID)
idx <- match(e12.5_pstime$tsne_v1, mouseE12.5_NMF_umap$annot$tsne_v1)

e12.5_pstime$UMAP1 <- mouseE12.5_NMF_umap$annot$UMAP1[idx]
e12.5_pstime$UMAP2 <- mouseE12.5_NMF_umap$annot$UMAP2[idx]

#e12.5_pstime <- mouseE12.5_NMF_umap$annot
```



```{r helper_functions}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                             Helper functions                               ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# Basic function to convert human to mouse gene names
# convertHumanGeneList <- function(x){
#   
#   human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
#   mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
#   
#   genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)
#   humanx <- unique(genesV2[, 2])
#   return(humanx)
# }

# Mouse to Human symbols
mm_hs_map = read.delim("http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt", stringsAsFactors = F)
mm_hs_map = mm_hs_map[,colnames(mm_hs_map) %in% c("HomoloGene.ID", "Symbol")]
mm_hs_map$HomoloGene.ID = as.character(mm_hs_map$HomoloGene.ID)
mm_hs_map = split(mm_hs_map$Symbol, mm_hs_map$HomoloGene.ID)
mm_hs_map = mm_hs_map[sapply(mm_hs_map, length) == 2] # Keeping only one to one homologs
mm_hs_map = do.call("rbind", mm_hs_map)
mm_hs_map = data.frame(mouse_symbol = noquote(as.character(mm_hs_map[,1])),
                       human_symbol = noquote(as.character(mm_hs_map[,2])))


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                        Extract associated features                         ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

extract_W_features <- function(wmatrix, omics){
  sigIDs <- setNames(colnames(wmatrix), colnames(wmatrix))
  
  lapply(sigIDs, function(sigID){
    # Extract signature specific genes
    selec_wmatrix <- do.call(cbind, lapply(as.data.frame(wmatrix), function(sign_expo){
      sign_expo[sign_expo < quantile(sign_expo, 0.90)] <- NA
      sign_expo
    }))
    rownames(selec_wmatrix) <- rownames(wmatrix)
    selec_wmatrix <- selec_wmatrix[!is.na(selec_wmatrix[,sigID]),,drop=FALSE]
    # Keep only the top feature if there's an overlap
    sig_SE_IDs <- rownames(selec_wmatrix[rowMaxs(selec_wmatrix, na.rm = TRUE) == selec_wmatrix[,sigID],])
    
    if (omics == "ChIPseq") {
      unique(SE[sig_SE_IDs]$target_SYMBOL)
    } else if (omics == "RNAseq") {
      unique(SE$target_SYMBOL[match(sig_SE_IDs, SE$target_ENSEMBL)])
    }
    
    #sign_expo  <- w_tumor[,sigID]
    #sig_SE_IDs <- names(sign_expo[sign_expo > quantile(sign_expo, 0.8)])
    
  })
}

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                        Projection plot                                     ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

projection_plot <- function(signature_associated, sigID){
  sig_geneIDs <- signature_associated[[sigID]]
  # Find Mouse Homologous
  #mouseIDs    <- convertHumanGeneList(sig_geneIDs)
  mouseIDs <- mm_hs_map$mouse_symbol[mm_hs_map$human_symbol %in% sig_geneIDs]
  # get mean expression
  geneSet_MeanZ <- colMeans(e12.5_exprs[rownames(e12.5_exprs) %in% mouseIDs,])
  # map to cells
  e12.5_pstime$exprsZ <- geneSet_MeanZ[match(e12.5_pstime$SampleID, names(geneSet_MeanZ))]
  # print plot
  e12.5_pstime %>% 
    #ggplot(aes(x = tsne_v1, y = tsne_v2, color = exprsZ)) +
    ggplot(aes(x = UMAP1, y = UMAP2, color = exprsZ)) +
    geom_point() +
    scale_color_gradient2()
  
  e12.5_pstime %>% 
    #ggplot(aes(x = UMAP1, y = UMAP2, color = exprsZ)) +
    #ggplot(aes(x = tsne_v1, y = tsne_v2, fill = exprsZ)) +
    ggplot(aes(x = UMAP1, y = UMAP2, fill = exprsZ)) +
    geom_point(size = 3, shape = 21, color = "grey50", stroke= 0.1) +
    #scale_fill_gradient2(low = "#0CD6F5", mid = "white", high = "#A8105A", labels=c("","low","","","high")) +
    scale_fill_gradient2(low = "#0CD6F5", mid = "white", high = "#A8105A",breaks = trans_breaks(identity, identity, n = 3)) +
    ggtitle(sigID) +
    ylab("Dimension 2") + 
    xlab("Dimension 1") + 
    theme_bw() +
    theme(legend.position = c(0.8, 0.1),
          legend.direction = "horizontal",
          legend.title = element_blank(),
          plot.title = element_text(hjust = 0.5),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank())
  
}
```

```{r extractFeatures}

signature_associated_ChIPseq <- extract_W_features(wc_tumor, omics = "ChIPseq")
signature_associated_RNAseq  <- extract_W_features(wr_tumor, omics = "RNAseq")



sigIDs <- setNames(colnames(wc_tumor), colnames(wc_tumor))

signature_associated <- lapply(sigIDs, function(sigID){
  chip <- signature_associated_ChIPseq[[sigID]]
  rna <- signature_associated_RNAseq[[sigID]]
  
  unique(c(chip, rna))
})



```



```{r ChIPseq_features}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          Plot projections                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

#sigIDs <- setNames(colnames(w_tumor), colnames(w_tumor))

mycn   <- projection_plot(signature_associated_ChIPseq, "MYCN")
mes    <- projection_plot(signature_associated_ChIPseq, "MES")
man_lr <- projection_plot(signature_associated_ChIPseq, "MNA-LR")
man_hr <- projection_plot(signature_associated_ChIPseq, "MNA-HR")


projection_gg <- mycn + mes + man_lr + man_hr + plot_layout(ncol = 2)
projection_gg


```


```{r RNAseq_features}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          Plot projections                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

#sigIDs <- setNames(colnames(w_tumor), colnames(w_tumor))

mycn   <- projection_plot(signature_associated_RNAseq, "MYCN")
mes    <- projection_plot(signature_associated_RNAseq, "MES")
man_lr <- projection_plot(signature_associated_RNAseq, "MNA-LR")
man_hr <- projection_plot(signature_associated_RNAseq, "MNA-HR")


projection_gg <- mycn + mes + man_lr + man_hr + plot_layout(ncol = 2)
projection_gg

ggsave(params$figure, projection_gg, width = 10, height = 10)

```



```{r ChIP_RNAseq_features}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          Plot projections                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

#sigIDs <- setNames(colnames(w_tumor), colnames(w_tumor))

mycn   <- projection_plot(signature_associated, "MYCN")
mes    <- projection_plot(signature_associated, "MES")
man_lr <- projection_plot(signature_associated, "MNA-LR")
man_hr <- projection_plot(signature_associated, "MNA-HR")


projection_gg <- mycn + mes + man_lr + man_hr + plot_layout(ncol = 2)
projection_gg


```

