---
title: "Figure 2h - expression mapped to Mouse GSE99933 E12.5"
author: "Health Data Science Unit"
date: "`r date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: false
params:
  SE_target    : "path/to/data"
  w_tumor      : "path/to/data"
  f_tumor      : "path/to/data"
  mouse_pstime : "path/to/data"
  mouse_exprs  : "path/to/data"
  figure : "results/figure..."
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
#this chunk sets global options for Rmarkdown
knitr::opts_chunk$set(echo = TRUE)
#setting working directory. Should be home directory of the project
options(knitr.table.format="html")

library(tidyverse)
library(viridis)
library(ggplot2)
library(biomaRt)
library(scales)
library(patchwork)
library(rtracklayer)

```



## Data loading
  
  
```{r dataloading}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                             Read data                                      ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# params = list(SE_target = "analysis/tumor/SE_annot/tumor_consensusSE_target_GRanges.RDS",
#               w_tumor = 'analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS',
#               f_tumor      = 'analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_NMF_features.RDS',
#               mouse_pstime = 'db/GSE99933_E12.5/GSE99933_E12.5.txt',
#               mouse_exprs  = 'db/GSE99933_E12.5/GSE99933_E12.5_exprs_Zscore.txt',
#               figure = 'results/figure2/figure2h_SEtargetExprs_mouseE12.5.pdf')

# Read SE annotation
SE <- readRDS(params$SE_target)

# Read NMF results
w_tumor <- readRDS(params$w_tumor)
f_tumor <- readRDS(params$f_tumor)


# Read Mouse GSE99933 E12.5 data
# Pseudotime and tsne axis
e12.5_pstime <- read.table(params$mouse_pstime, 
                           header = TRUE,
                           stringsAsFactors = FALSE)

head(e12.5_pstime)
# Zscore of gene expression
e12.5_exprs <- read.table(params$mouse_exprs, 
                          header = TRUE,
                          colClasses = c(probeset = "NULL"),
                          row.names = 1,
                          stringsAsFactors = FALSE)
e12.5_exprs[1:5,1:5]




```



```{r}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                             Helper functions                               ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# Basic function to convert human to mouse gene names
convertHumanGeneList <- function(x){
  
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
  
  genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)
  humanx <- unique(genesV2[, 2])
  return(humanx)
}


# Projections
projection_plot <- function(sigID){
  # Extract signature specific gens
  sig_SE_IDs  <- rownames(f_tumor[rowSums(f_tumor) == 1 & f_tumor[, sigID] == 1, ])
  sig_geneIDs <- unique(SE[sig_SE_IDs]$target_SYMBOL)
  # Find Mouse Homologous
  mouseIDs    <- convertHumanGeneList(sig_geneIDs)
  # get mean expression
  geneSet_MeanZ <- colMeans(e12.5_exprs[rownames(e12.5_exprs) %in% mouseIDs,])
  # map to cells
  e12.5_pstime$exprsZ <- geneSet_MeanZ[match(e12.5_pstime$SampleID, names(geneSet_MeanZ))]
  # print plot
  e12.5_pstime %>% 
    ggplot(aes(x = tsne_v1, y = tsne_v2, color = exprsZ)) +
    geom_point() +
    scale_color_gradient2()
  
  e12.5_pstime %>% 
    ggplot(aes(x = tsne_v1, y = tsne_v2, fill = exprsZ)) +
    geom_point(size = 3, shape = 21, color = "grey50", stroke= 0.1) +
    #scale_fill_gradient2(low = "#0CD6F5", mid = "white", high = "#A8105A", labels=c("","low","","","high")) +
    scale_fill_gradient2(low = "#0CD6F5", mid = "white", high = "#A8105A",breaks = trans_breaks(identity, identity, n = 3)) +
    ggtitle(sigID) +
    ylab("Dimension 2") + 
    xlab("Dimension 1") + 
    theme_bw() +
    theme(legend.position = c(0.8, 0.1),
          legend.direction = "horizontal",
          legend.title = element_blank(),
          plot.title = element_text(hjust = 0.5),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank())
  
}



##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          Plot projections                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

sigIDs <- setNames(colnames(f_tumor), colnames(f_tumor))

mycn   <- projection_plot("MYCN")
mes    <- projection_plot("MES")
man_lr <- projection_plot("MNA-LR")
man_hr <- projection_plot("MNA-HR")


projection_gg <- mycn + mes + man_lr + man_hr + plot_layout(ncol = 2)
projection_gg

ggsave(params$figure, projection_gg, width = 10, height = 10)


```

```{r}
projection_plot <- function(sigID){
  # Extract signature specific gens
  sign_expo  <- w_tumor[,sigID]
  sig_SE_IDs <- names(sign_expo[sign_expo > quantile(sign_expo, 0.8)])
  sig_geneIDs <- unique(SE[sig_SE_IDs]$target_SYMBOL)
  # Find Mouse Homologous
  mouseIDs    <- convertHumanGeneList(sig_geneIDs)
  # get mean expression
  geneSet_MeanZ <- colMeans(e12.5_exprs[rownames(e12.5_exprs) %in% mouseIDs,])
  # map to cells
  e12.5_pstime$exprsZ <- geneSet_MeanZ[match(e12.5_pstime$SampleID, names(geneSet_MeanZ))]
  # print plot
  e12.5_pstime %>% 
    ggplot(aes(x = tsne_v1, y = tsne_v2, color = exprsZ)) +
    geom_point() +
    scale_color_gradient2()
  
  e12.5_pstime %>% 
    ggplot(aes(x = tsne_v1, y = tsne_v2, fill = exprsZ)) +
    geom_point(size = 3, shape = 21, color = "grey50", stroke= 0.1) +
    #scale_fill_gradient2(low = "#0CD6F5", mid = "white", high = "#A8105A", labels=c("","low","","","high")) +
    scale_fill_gradient2(low = "#0CD6F5", mid = "white", high = "#A8105A",breaks = trans_breaks(identity, identity, n = 3)) +
    ggtitle(sigID) +
    ylab("Dimension 2") + 
    xlab("Dimension 1") + 
    theme_bw() +
    theme(legend.position = c(0.8, 0.1),
          legend.direction = "horizontal",
          legend.title = element_blank(),
          plot.title = element_text(hjust = 0.5),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_blank())
  
}


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          Plot projections                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

sigIDs <- setNames(colnames(f_tumor), colnames(f_tumor))

mycn   <- projection_plot("MYCN")
mes    <- projection_plot("MES")
man_lr <- projection_plot("MNA-LR")
man_hr <- projection_plot("MNA-HR")


projection_gg <- mycn + mes + man_lr + man_hr + plot_layout(ncol = 2)
projection_gg

ggsave(params$figure, projection_gg, width = 10, height = 10)

```

