---
title: "Supp. tables"
author: "Health Data Science Unit"
date: "`r date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: false
params:
  #annot_t    : "annotation/annotation_tumor.RDS"
  annot_t    : "annotation/annotation_tumor_with_external_phenoData.RDS"
  annot_c    : "annotation/annotation_cells.RDS"
  purity_t   : "annotation/purity_tumor.csv"
  SE         : "analysis/tumor/SE_annot/tumor_consensusSE_target_GRanges.RDS"
  go_enrich  : "results/supptables/GO_BP_enrichment_SE_target_genes.txt"
  eChIP_tumor : "analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_GO_BP_enrichment.RDS"
  eChIP_cells : "analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_GO_BP_enrichment.RDS"
  eRNAs_tumor : "analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_GO_BP_enrichment.RDS"
  eRNAv_tumor : "analysis/tumor/rnaseq/NMF_mostVariable/tumor_mostVariable_K4_GO_BP_enrichment.RDS"
  Hchip_t    : "analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Hmatrix_wnorm.RDS"
  Hchip_c    : "analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_Hmatrix_wnorm.RDS"
  Hchip_tc   : "analysis/tumor_cells/chipseq/H3K27ac/NMF/tumor_cells_consensusSE_K5_Hmatrix_wnorm.RDS"
  Hrna_tt    : "analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Hmatrix_wnorm.RDS"
  Hrna_tmv   : "analysis/tumor/rnaseq/NMF_mostVariable/tumor_mostVariable_K4_Hmatrix_wnorm.RDS"
  NBregulome : "analysis/tumor/ARACNe/network.txt"
  TFactivity : "results/supptables/TFactivity_across_all_signatures_ZnormPerSig.txt"
  CRCfracobs : "results/supptables/crcTF_fractionObserved.xls"
  CRCmodules : "results/supptables/crcTF_modules.txt"
  EnrichTF   : "analysis/tumor/Rel_vs_Pri/RelapseVsPrimary_EnrichTFregulons.txt"
  S1 : "results/supptables/S1_annotation.xlsx"
  S2 : "results/supptables/S2_SE_regions_and_target.xlsx"
  S3 : "results/supptables/S3_SE_target_genes_GO_BP_enrichment.xlsx"
  S4 : "results/supptables/S4_NMF_H_Matrices.xlsx"
  S5 : "results/supptables/S5_NB_regulome.xlsx"
  S6 : "results/supptables/S6_TFactivity_across_all_signatures.xlsx"
  S7 : "results/supptables/S7_crcTF_fractionObserved.xlsx"
  S8 : "results/supptables/S8_RelapseVsPrimary_EnrichTFregulons.xlsx"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
#this chunk sets global options for Rmarkdown
knitr::opts_chunk$set(echo = TRUE)
#setting working directory. Should be home directory of the project
knitr::opts_knit$set(root.dir=params$work_dir)
options(knitr.table.format="html")

library(GenomicRanges)
library(writexl)
library(readxl)
library(tidyverse)
library(knitr)


```

# Compile supp tables


```{r dataloading}

# params <- list(annot_t   = "annotation/annotation_tumor.RDS",
#                annot_c   = "annotation/annotation_cells.RDS",
#                purity_t  = "annotation/purity_tumor.csv",
#                SE        = "analysis/tumor/SE_annot/tumor_consensusSE_target_GRanges.RDS",
#                go_enrich = "results/supptables/GO_BP_enrichment_SE_target_genes.txt",
# 
#                Hchip_t   = "analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Hmatrix_wnorm.RDS",
#                Hchip_c   = "analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_Hmatrix_wnorm.RDS",
#                Hchip_tc  = "analysis/tumor_cells/chipseq/H3K27ac/NMF/tumor_cells_consensusSE_K5_Hmatrix_wnorm.RDS",
#                Hrna_tt   = "analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Hmatrix_wnorm.RDS",
#                Hrna_tmv  = "analysis/tumor/rnaseq/NMF_mostVariable/tumor_mostVariable_K4_Hmatrix_wnorm.RDS",
# 
#                NBregulome = "analysis/tumor/ARACNe/network.txt",
#                TFactivity = "results/supptables/TFactivity_across_all_signatures_ZnormPerSig.txt",
#                CRCfracobs = "results/supptables/crcTF_fractionObserved.xls",
#                CRCmodules = "results/supptables/crcTF_modules.txt",
#                EnrichTF   = "analysis/tumor/Rel_vs_Pri/RelapseVsPrimary_EnrichTFregulons.txt",
# 
#                S1 = "results/supptables/S1_annotation.xlsx",
#                S2 = "results/supptables/S2_SE_regions_and_target.xlsx",
#                S3 = "results/supptables/S3_SE_target_genes_GO_BP_enrichment.xlsx",
#                S4 = "results/supptables/S4_NMF_H_Matrices.xlsx",
#                S5 = "results/supptables/S5_NB_regulome.xlsx",
#                S6 = "results/supptables/S6_TFactivity_across_all_signatures.xlsx",
#                S7 = "results/supptables/S7_crcTF_fractionObserved.xlsx",
#                S8 = "results/supptables/S8_RelapseVsPrimary_EnrichTFregulons.xlsx")

# add rownames as frist column

paste_rownames <- function(x, colID){
  idcol <- data.frame(rownames(x))
  colnames(idcol) <- colID
  data.frame(idcol, x)
}


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          S1_annotation                                     ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

#annot_t <- readRDS(params$annot_t)
annot_t <- readRDS("annotation/annotation_tumor_with_external_phenoData.RDS")
annot_c <- readRDS(params$annot_c)
purity_t <- read.csv(params$purity_t)

annot_t$Purity <- purity_t$Purity[match(annot_t$ProjectID, purity_t$ProjectID)]
annot_t$Ploidy <- purity_t$Ploidy[match(annot_t$ProjectID, purity_t$ProjectID)]
colnames(annot_t)


# Add signature assignment 

tumor_hchip <- readRDS("analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS")
tumor_hchip <- setNames(rownames(tumor_hchip)[apply(tumor_hchip, 2, which.max)],
                             colnames(tumor_hchip))

tumor_hrna <- readRDS("analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS")
tumor_hrna <- setNames(rownames(tumor_hrna)[apply(tumor_hrna, 2, which.max)],
                             colnames(tumor_hrna))

annot_t <- annot_t %>% 
  mutate(Signature_ChIPseq = tumor_hchip[match(ProjectID, names(tumor_hchip))]) %>% 
  mutate(Signature_RNAseq = tumor_hrna[match(ProjectID, names(tumor_hrna))])

# Save 
saveRDS(annot_t, "annotation/annotation_tumor_phenoData_NMFassign.RDS")

#lapply(list, function)


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          Assignmet confidence                              ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##


sigmoidTransform <- function(col.vector, q = 0.95) {
  q <- as.numeric(quantile(col.vector, q))
  x <- 2 / (1 + exp((-2) * col.vector / q)) - 1
  return(x)
}



Hchip_t  <- readRDS("analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS")
Hrna_tt  <- readRDS("analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS")


best_assign <- function(H, iter){
  
  res <- lapply(1:iter, function(i){
    apply(H, 1, function(x){
      #x <- sigmoidTransform(x)
      #x <- x/max(x)
      k <- kmeans(x, 2)
      max_idx <- which.max(k$centers)
      setNames(if_else(k$cluster == max_idx, 1, 0), names(x))
      })
  })
  #print(lapply(res, function(i) table(rowSums(i))))
  best <- sapply(res, function(i){sum(rowSums(i) == 1)})
  res <- res[[which.max(best)]]
  #res <- res[[which.min(best)]]
  
  setNames(if_else(rowSums(res) == 1, "High confidence", "Low confidence"),
           colnames(H))
}

Hchip_conf <- best_assign(H = Hchip_t, iter = 1000)
table(Hchip_conf)
Hchip_conf


Hrna_conf <- best_assign(H = Hrna_tt, iter = 1000)
table(Hrna_conf)
Hrna_conf


Hchip_t  <- as.data.frame(t(readRDS("analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS"))) %>% 
  #Hchip_t  <- as.data.frame(t(readRDS(params$Hchip_t))) %>% 
  rename_all(paste0, "_ChIP-seq") %>% 
  rownames_to_column("ProjectID") 
Hrna_tt  <- as.data.frame(t(readRDS("analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS"))) %>% 
  #Hrna_tt  <- as.data.frame(t(readRDS(params$Hrna_tt))) %>% 
  rename_all(paste0, "_RNA-seq") %>% 
  rownames_to_column("ProjectID") 

#annot_t %>% 
annot_t <- annot_t %>% 
  mutate(Signature_ChIPseq_confidence = Hchip_conf[match(ProjectID, names(Hchip_conf))]) %>% 
  mutate(Signature_RNAseq_confidence = Hrna_conf[match(ProjectID, names(Hrna_conf))]) %>% 
  mutate(Signature_ChIPseq_confidence = if_else(is.na(Signature_ChIPseq_confidence), "na", Signature_ChIPseq_confidence)) %>% 
  mutate(Signature_RNAseq_confidence = if_else(is.na(Signature_RNAseq_confidence), "na", Signature_RNAseq_confidence)) %>% 
  mutate(Signature_ChIPseq = if_else(Signature_ChIPseq_confidence == "High confidence", paste0("_hccc_", Signature_ChIPseq), Signature_ChIPseq)) %>% 
  mutate(Signature_RNAseq = if_else(Signature_RNAseq_confidence == "High confidence", paste0("_hccc_", Signature_RNAseq), Signature_RNAseq)) %>% 
  mutate(Signature_ChIPseq = if_else(is.na(Signature_ChIPseq), "_lqqq_", Signature_ChIPseq)) %>% 
  mutate(Signature_RNAseq = if_else(is.na(Signature_RNAseq), "_lqqq_", Signature_RNAseq)) %>% 
  
  left_join(Hchip_t, by = "ProjectID") %>% 
  left_join(Hrna_tt, by = "ProjectID")


annot_t %>% 
  filter(avail.RNAseq & avail.ChIPseq) %>% 
  filter(Signature_ChIPseq != Signature_RNAseq) %>% 
  filter(Signature_ChIPseq_confidence == "High confidence") %>% 
  filter(Signature_RNAseq_confidence == "High confidence") 
  




write_xlsx(list(Tumors       = annot_t,
                `Cell lines` = annot_c), 
           path = params$S1)


annot_t %>% 
  filter(avail.RNAseq & avail.ChIPseq) %>% 
  filter(Signature_ChIPseq != Signature_RNAseq) %>% 
  group_by(Signature_ChIPseq_confidence) %>% 
  summarise(Assign_ChIPseq_confidence = n())


annot_t %>% 
  filter(avail.RNAseq & avail.ChIPseq) %>% 
  filter(Signature_ChIPseq != Signature_RNAseq) %>% 
  group_by(Signature_RNAseq_confidence) %>% 
  summarise(Assign_RNAseq_confidence = n())



```

```{r s2}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                    S2_SE_regions_and_target                                ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

SE <- readRDS(params$SE)
SE <- as.data.frame(SE)
SE <- paste_rownames(SE, "SE_ID")

# Add rank
SE_total_Signal_by_ID <- readRDS("analysis/tumor/chipseq/H3K27ac/consensusSE/tumor_H3K27ac_noH3K4me3_SE_total_Signal.RDS")
SE_rank <- setNames(rowMedians(SE_total_Signal_by_ID), rownames(SE_total_Signal_by_ID))
#SE$SE_rank <- SE_rank[match(SE$SE_ID, names(SE_rank))]
SE$SE_rank <- rank(SE_rank[match(SE$SE_ID, names(SE_rank))])



#------------------------------------------------------------------------------#
#                              Add SE assignment                               #
#------------------------------------------------------------------------------#
# Add signature assignment 
tumor_hchip <- readRDS("analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS")
tumor_hrna  <- readRDS("analysis/tumor/rnaseq/NMF/tumor_consensusSE_K4_Wmatrix_Wnorm.RDS")


#------------------------------------------------------------------------------#
#                  Add sig assignment based on top 10%                         #
#------------------------------------------------------------------------------#
top_10perc_assing <- function(wmatrix){
  sig_assign <- lapply(setNames(colnames(wmatrix), colnames(wmatrix)), function(sigID){
    selec_wmatrix <- do.call(cbind, lapply(as.data.frame(wmatrix), function(sign_expo){
      sign_expo[sign_expo < quantile(sign_expo, 0.90)] <- NA
      sign_expo
    }))
    rownames(selec_wmatrix) <- rownames(wmatrix)
    selec_wmatrix <- selec_wmatrix[!is.na(selec_wmatrix[,sigID]),,drop=FALSE]
    # Keep only the top feature if there's an overlap
    sig_SE_IDs <- rownames(selec_wmatrix[rowMaxs(selec_wmatrix, na.rm = TRUE) == selec_wmatrix[,sigID],])
    setNames(rep(sigID, length(sig_SE_IDs)), sig_SE_IDs)
  })
  unlist(unname(sig_assign))
}


tumor_hchip_top10perc <- top_10perc_assing(tumor_hchip)
tumor_hrna_top10perc <- top_10perc_assing(tumor_hrna)


#------------------------------------------------------------------------------#
#                  Add sig assignment based on cluster                         #
#------------------------------------------------------------------------------#

# Run k means over all rows and assign features to the clusters
SignFeatures <- function(wmatrix){
  ssf <- apply(wmatrix, 1, function(x){
    #print(x)
    x <- sigmoidTransform(x)
    k <- kmeans(x, 2)
    max_idx <- which.max(k$centers)
    #paste(if_else(k$cluster == max_idx, "1", "0"), collapse = "")
    setNames(if_else(k$cluster == max_idx, 1, 0), names(x))
  })
  ssf
  
  ssf <- ssf[,colSums(ssf) == 1]
  setNames(rownames(ssf)[apply(ssf, 2, which.max)], colnames(ssf))
}


# Add signature assignment 
tumor_hchip_kmeansk2 <- SignFeatures(tumor_hchip)
tumor_hrna_kmeansk2 <- SignFeatures(tumor_hrna)


#------------------------------------------------------------------------------#
#                  Add sig assignment based max exposure                       #
#------------------------------------------------------------------------------#

tumor_hchip_maxexposure <- setNames(colnames(tumor_hchip)[apply(tumor_hchip, 1, which.max)],
                                    rownames(tumor_hchip))
tumor_hrna_maxexposure <- setNames(colnames(tumor_hrna)[apply(tumor_hrna, 1, which.max)],
                                    rownames(tumor_hrna))




head(SE)
SE <- SE %>% 
  mutate(Signature_ChIPseq_Top10Percent = tumor_hchip_top10perc[match(SE_ID, names(tumor_hchip_top10perc))]) %>% 
  mutate(Signature_ChIPseq_KmeansK2     = tumor_hchip_kmeansk2[match(SE_ID, names(tumor_hchip_kmeansk2))]) %>% 
  mutate(Signature_ChIPseq_MaxExposure  = tumor_hchip_maxexposure[match(SE_ID, names(tumor_hchip_maxexposure))]) %>% 
  
  mutate(Signature_Target_RNAseq_Top10Percent = tumor_hrna_top10perc[match(target_ENSEMBL, names(tumor_hrna_top10perc))]) %>% 
  mutate(Signature_Target_RNAseq_KmeansK2     = tumor_hrna_kmeansk2[match(target_ENSEMBL, names(tumor_hrna_kmeansk2))]) %>% 
  mutate(Signature_Target_RNAseq_MaxExposure  = tumor_hrna_maxexposure[match(target_ENSEMBL, names(tumor_hrna_maxexposure))])

table(SE$Signature_ChIPseq_MaxExposure, SE$Signature_Target_RNAseq_MaxExposure)
table(SE$Signature_ChIPseq_Top10Percent, SE$Signature_Target_RNAseq_Top10Percent)
table(SE$Signature_ChIPseq_KmeansK2, SE$Signature_Target_RNAseq_KmeansK2)


saveRDS(SE, "analysis/tumor/SE_annot/tumor_consensusSE_NMF_assignment.RDS")


SE <- SE %>% 
  select(-Signature_ChIPseq_KmeansK2) %>% 
  select(-Signature_Target_RNAseq_KmeansK2) #%>% 
  #rename(Signature_ChIPseq_Top10Percent = "Signature_ChIPseq")
 #Signature_ChIPseq_MaxExposure

write_xlsx(list(`Super Enhancers` = SE), 
           path = params$S2)

```



```{r}


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                S3_SE_target_genes_GO_BP_enrichment                         ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

go_enrich <- read.table(params$go_enrich, sep = "\t", header = TRUE)
eChIP_tumor <- readRDS(params$eChIP_tumor)
eChIP_cells <- readRDS(params$eChIP_cells)
eRNAs_tumor <- readRDS(params$eRNAs_tumor)
eRNAv_tumor <- readRDS(params$eRNAv_tumor)

enrich_list <- list(`GO BP enrichment` = go_enrich,
                    `Tumor SE H3K27ac` = eChIP_tumor,
                    `Cell lines SE H3K27ac` = eChIP_cells,
                    `Tumor SE targets RNAseq` = eRNAs_tumor,
                    `Tumor RNAseq most variable` = eRNAv_tumor)


write_xlsx(enrich_list, 
           path = params$S3)


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                       S4_NMF_H_Matrices                                    ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
Hchip_t  <- readRDS(params$Hchip_t)
Hchip_c  <- readRDS(params$Hchip_c)
Hchip_tc <- readRDS(params$Hchip_tc)
Hrna_tt  <- readRDS(params$Hrna_tt)
Hrna_tmv <- readRDS(params$Hrna_tmv)

h_list <- list(`Tumor SE H3K27ac` = Hchip_t,
               `Cell lines SE H3K27ac` = Hchip_c,
               `Tumor & Cell lines SE H3K27ac` = Hchip_tc,
               `Tumor SE targets RNAseq` = Hrna_tt,
               `Tumor RNAseq most variable` = Hrna_tmv)

h_list <- lapply(h_list, as.data.frame)
h_list <- lapply(h_list, paste_rownames, colID = "SignatureID")

write_xlsx(h_list, 
           path = params$S4)




##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                       S5_NB_regulome                                       ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
NBregulome  <- read.table(params$NBregulome, header = TRUE)

write_xlsx(list(`NB regulome` = NBregulome), 
           path = params$S5)


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                 S6_TFactivity_across_all_signatures                        ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
TFactivity  <- read.table(params$TFactivity, header = TRUE)

write_xlsx(list(`TF activity` = TFactivity), 
           path = params$S6)


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                 S7_crcTF_fractionObserved                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

CRCfracobs <- list(`Sig combined` = read_excel(params$CRCfracobs, sheet = 1),
                   `Sig separate` = read_excel(params$CRCfracobs, sheet = 2),
                   modules        = read.table(params$CRCmodules, header = TRUE))

write_xlsx(CRCfracobs, 
           path = params$S7)


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                 S8_RelapseVsPrimary_EnrichTFregulons                       ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

EnrichTF <- list(`Enrich TF regulons` = read.table(params$EnrichTF, header = TRUE, sep = "\t"))

write_xlsx(EnrichTF, 
           path = params$S8)


```

