---
title: "Enhancer comparison plot"
author: "Health Data Science Unit"
date: "`r date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: false
params:
  work_dir     : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/"
  SE           : "analysis/tumor/chipseq/H3K27ac/consensusSE/tumor_H3K27ac_noH3K4me3_consensusSE.bed"
  cells_h      : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_Hmatrix_hnorm.RDS"
  cells_w      : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_Wmatrix_Wnorm.RDS"
  tumor_h      : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS"
  SKNAS_HiChIP : "data/cells/hichip/mango/SK-N-AS_HiChIP_mango.all"
  CLBGA_HiChIP : "data/cells/hichip/mango/CLB-GA_HiChIP_mango.all" 
  figure       : "results/figure3/Super_Enhancer_interaction.pdf"
  sup_figure   : "results/figure3/Enhancer_interaction.pdf"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Differential interaction analysis between enhacner and super enhancer elements as captured with HiChIP. Datat from cell lines SK-N-AS and CLB-GA pulled down on H3K27ac.

```{r echo=F, message=FALSE, warning=F}
#setwd("/home/saary/hichip/data/20170925-SEcomparison")

#load('datatforplot.Rda')

library(GenomicInteractions)
library(ggplot2)
library(cowplot)
library(tidyverse)
library(reshape2)
library(DiffBind)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

# fetch enhancers for mesochymal of non mechchimal samples
ColMes     <- "#40004B"
ColNo      <- "#006837"

```

## Custom functions for analysis

```{r message=FALSE, warning=FALSE, error = FALSE}


makeSE <- function(x){
  GRanges(x[,1], IRanges(x[,2], x[,3]))
}

makeInteraction <- function(x){
  gr <- GRanges(x[,1], IRanges(x[,2], x[,3]))
  gr@elementMetadata$name <- x[,4]
  return(gr)
}


rarefy <- function(data, m){
  # a simple rarefaction or downsampling function
  if(m > length(data)){
    stop("m > data")
  }
  s <- sample(seq(1:length(data)), m, replace = F)
  return(data[s])
}


# load both datasets
readGI <- function(x){
  m <- read.table(x)
  I <- GenomicInteractions(GRanges(m[,1], IRanges(m[,2],m[,3])),GRanges(m[,4], IRanges(m[,5],m[,6])))
  return(I)
}

load.bed <- function(file){
  makeInteraction(read.table(file))
}

write.bed <- function(gr, filename, ...){
  df <- data.frame(seqnames=seqnames(gr),
                   starts=start(gr)-1,
                   ends=end(gr),
                   names=c(paste("enhancer", seq(1, length(gr)), sep = "_")),
                   strands=strand(gr))
  
  write.table(df, file=filename, quote=F, sep="\t", row.names=F, col.names=F, ...)
}

# find regions that have interactions and are mesonchymal or not
filterRegions <- function(bed, set){
  # return regions that have at least one interaction in them
  subsetBed <- bed[overlapsAny(bed, set),]
  return(subsetBed)
}

compareBoth <- function(x, S, C, enhancer, m){
  r <- rep(NA, 8)
  names(r) <- c("Smc", "Snmc","Sunkown","Snone", "Cmc", "Cnmc","Cunkown", "Cnone")
  
  S <- rarefy(S,m)
  C <- rarefy(C,m)
  
  se <- enhancer[overlapsAny(enhancer,S),]
  r[1] <- length(se[se@elementMetadata$Fold < 0,])
  r[2] <- length(se[se@elementMetadata$Fold > 0,])
  r[3] <- length(se[se@elementMetadata$Fold == 0,])
  r[4] <- length(S) - length(se)
  ce <- enhancer[overlapsAny(enhancer, C),]
  r[5] <- length(ce[ce@elementMetadata$Fold < 0,])
  r[6] <- length(ce[ce@elementMetadata$Fold > 0,])
  r[7] <- length(ce[ce@elementMetadata$Fold == 0,])
  r[8] <- length(C) - length(ce)
 
  return(r)
  
}

analyseInteractions <- function(regions, setA, setB, n, m){
  ps     <- sapply(1:n, compareBoth, setA,setB, regions, m)
  ps     <- t(ps)/m
  a <- as.data.frame(ps[,1:4])
  b <- as.data.frame(ps[,5:8])
  colnames(a) <- c("mesenchymal", "non mesenchymal", "not classified", "not in enhancer")
  colnames(b) <- c("mesenchymal", "non mesenchymal", "not classified", "not in enhancer")
  
  a$type      <- "SK-N-AS"
  b$type      <- "CLB-GA"
  
  df          <- melt(rbind(a,b), id.vars = "type")
  return(df)
}

```

## Diffbind analysis


Do diffbind analysis using bam files and MACS2 called peaks of H3K27ac ChIP-seq.
```{r DIFFBIND, message=FALSE, warning=FALSE, cache=FALSE}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                     assign samples to MES signature                        ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
assign_signature <- function(h){
  is_h <- t(apply(h, 1, function(x){
    k <- kmeans(x, 2)
    max_idx <- which.max(k$centers)
    if_else(k$cluster == max_idx, TRUE, FALSE)
  }))
  colnames(is_h) <- colnames(h)
  rownames(is_h) <- rownames(h)
  is_h
}

cells_h <- readRDS(params$cells_h)
tumor_h <- readRDS(params$tumor_h)

cells_mes <- assign_signature(cells_h)["MES",]
tumor_mes <- assign_signature(tumor_h)["MES",]

cells_mes
tumor_mes

table(cells_mes)
table(tumor_mes)

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                                  DIFFBIND                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
tumor_bam <- list.files(file.path(params$work_dir, "data/tumor/chipseq/H3K27ac/bam"), 
                        pattern = "_H3K27ac_rmdup_sorted_merged.bam", full.names = TRUE)
message(tumor_bam)
cells_bam <- list.files(file.path(params$work_dir, "data/cells/chipseq/H3K27ac/bam"), 
                        pattern = "_H3K27ac_rmdup_sorted_merged.bam", full.names = TRUE)



samples <- rbind(data.frame(SampleID = sub("_H3K27ac_rmdup_sorted_merged.bam", "", basename(tumor_bam)),
                                 bamReads = tumor_bam,
                                 Factor   = "tissue"),
                      data.frame(SampleID = sub("_H3K27ac_rmdup_sorted_merged.bam", "", basename(cells_bam)),
                                 bamReads = cells_bam,
                                 Factor   = "cell line"))
samples


mes_sampleIDs <- c(names(cells_mes)[cells_mes == TRUE], names(tumor_mes)[tumor_mes == TRUE])
samples$Condition  <- "ADRN"
samples[samples$SampleID %in% mes_sampleIDs,]$Condition <- "MES"



# # load chipseq files for Diffbind analysis
# tissue             <- read.table('/home/saary/hichip/data/20170925-SEcomparison/bam-tissue.csv', stringsAsFactors = F)
# colnames(tissue)   <- c("SampleID",  "bamReads")
# tissue$Factor      <- "tissue"
# 
# cellline           <- read.table('/home/saary/hichip/data/20170925-SEcomparison/bam-cellline.csv', stringsAsFactors = F)
# colnames(cellline) <- c("SampleID",  "bamReads")
# cellline$Factor    <- "cell line"
# 
# # combine data of cell lines and tumor samples
# samples            <- rbind(tissue, cellline)
# 
# 
# # define mesenchymal samples/tumors
# meso               <- c("B087_HD-N-33", "SK-N-AS", "B087_SH-EP", "P5", "P4", "GI-ME-N", "NBS325", "B087_N735", "B087_N699")
# samples$Condition  <- "non mesenchymal"
# samples[samples$SampleID %in% meso,]$Condition <- "mesenchymal"

# SE
# consensusSE <- readRDS(params$SE)
# 
# peaksD <- as.data.frame(unique(consensusSE))[,1:3]
# peaksD[,4] <- rownames(peaksD)
# peaksF <- c("/home/saary/hichip/data/20170925-SEcomparison/E_filtered_score.bed")
# write.table(x = peaksD, row.names = F, file = peaksF, col.names = F, sep = "\t", quote = F)

samples$Peaks      <- params$SE
samples$PeakCaller <- "bed"


# do the diffbind analysis
d    <- dba(sampleSheet = samples)
d    <- dba.count(d)
d    <- dba.contrast(d, categories = "Condition", minMembers = 2)
d    <- dba.analyze(d)
d.db <- dba.report(d)

# load the consensus peaks
p    <- load.bed(samples$Peaks[1])

# split in two sets, for later use

Emc   <- d.db[(d.db$Conc_MES - d.db$Conc_ADRN) > 0,]
Enm   <- d.db[(d.db$Conc_MES - d.db$Conc_ADRN) < 0,]


allID <- seq(1:length(p))
b     <- findOverlaps(p, d.db, type="equal")@from
Enone <- p[allID[!allID %in% b],]
Enone@elementMetadata$name <- NULL
Enone@elementMetadata$Conc <- 0
Enone@elementMetadata$Conc_ADRN <- 0
Enone@elementMetadata$Conc_MES  <- 0
Enone@elementMetadata$Fold <- 0
Enone@elementMetadata$'p-value' <- 0
Enone@elementMetadata$FDR <- 0

# enhancers as set of mesenchymal, non mesenchymal and neither nor peaks
enhancer <- c(Emc, Enm, Enone)
### END DIFFBIND
####################################################
```


## Differential interaction calling

Interactions of HiChIP were called using Mango. After aligning the fastq files using Hi-C-pro , mango steps 4 and 5 were used to call interactions witha FDR treshold of 0.1.

```{r Interactions, cache = FALSE}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                    assign SE to MES signature                              ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
### Super enhancers
SE   <- load.bed(params$SE)

# Read W matrix and assign features to each signature
w_matrix <- readRDS(params$cells_w)
sign_feat <- lapply(setNames(colnames(w_matrix), colnames(w_matrix)), function(sigID){
  wmatrix <- w_matrix
  # Extract signature specific gens
  selec_wmatrix <- do.call(cbind, lapply(as.data.frame(wmatrix), function(sign_expo){
    sign_expo[sign_expo < quantile(sign_expo, 0.90)] <- NA
    sign_expo
  }))
  rownames(selec_wmatrix) <- rownames(wmatrix)
  selec_wmatrix <- selec_wmatrix[!is.na(selec_wmatrix[,sigID]),,drop=FALSE]
  # Keep only the top feature if there's an overlap
  sig_SE_IDs <- rownames(selec_wmatrix[rowMaxs(selec_wmatrix, na.rm = TRUE) == selec_wmatrix[,sigID],])
})


# add random score
SE$Fold <- 0

SE[SE$name %in% sign_feat$MES]$Fold  <- 1
SE[SE$name %in% sign_feat$MYCN]$Fold <- -1
SE[SE$name %in% sign_feat$MNA]$Fold  <- -1

# mc   <- read.table("/home/saary/analysis/nb/analysis/enhancerComparison/SE_specific_sig1.csv")
# nmc   <- read.table("/home/saary/analysis/nb/analysis/enhancerComparison/SE_specific_NonSig1.csv")
# MC < 0 non MC > 0
# SE[SE$name %in% mc$V1,]$Fold <- -1
# SE[SE$name %in% nmc$V1,]$Fold <- 1

# remove unspecific:
SE <- SE[SE$Fold != 0]

#SEmc   <- SE[which(SE@elementMetadata$Fold < 0),]
#SEnm   <- SE[which(SE@elementMetadata$Fold > 0),]
#SEnone <- SE[which(SE@elementMetadata$Fold == 0),]

superEnhancer <- SE
### End Superenhancer
####################################################



# load both HiChip bed files 
# S    <- readGI("/home/saary/hichip/data/20170925-SEcomparison/SK-N-AS-k27.bedI")
# C    <- readGI("/home/saary/hichip/data/20170925-SEcomparison/CLB-GA-k27.bedI")
S    <- readGI(params$SKNAS_HiChIP)
C    <- readGI(params$CLBGA_HiChIP)

# randomly sample from C and S, get minimal m
msize  <- min(c(length(S),length(C)))
n      <- 500
m      <- round(msize * 0.75)



# Enhancer
a   <- analyseInteractions(enhancer, S, C, n , m)
a$c <- "enhancer"
# SuperEnhancer
b   <- analyseInteractions(superEnhancer, S, C, n , m)
b$c <- "super enhancer"
df <- rbind(a, b)

```

```{R SE}
mesMes <- sum((countOverlaps(superEnhancer[superEnhancer$Fold == 1], S)))/length(S)
mesNon <- sum((countOverlaps(superEnhancer[superEnhancer$Fold == 1], C)))/length(C)
nonMes <- sum((countOverlaps(superEnhancer[superEnhancer$Fold == -1], S)))/length(S)
nonNon <- sum((countOverlaps(superEnhancer[superEnhancer$Fold == -1], C)))/length(C)
df <- data.frame(mm = mesMes,
           mn  = mesNon,
           nm = nonMes,
           nn = nonNon)
df2 <- data.frame(mm = mesMes*length(S),
           mn  = mesNon*length(C),
           nm = nonMes*length(S),
           nn = nonNon*length(C))

res <- c(mesMes, mesNon, nonMes, nonNon)
res <- res/sum(res)
names(res) <- c("mesSE SK-N-AS", "mesSE CLB-GA", "non SE SKNAS", "non CLBGA")
barplot(res, legend =names(res)  )

mSESum <- mesMes + mesNon
mm    <- mesMes/mSESum
mn    <- mesNon/mSESum
nSESum <- nonMes + nonNon
nm <- nonMes/nSESum
nn <- nonNon/nSESum

stacked <-c(mm,mn, nm,nn) 
names(stacked) <- c(rep("MES",2), rep("ADRN",2))
mat <- matrix(stacked,2)
colnames(mat) <- c("MES", "ADRN")
barplot(mat, legend = colnames(mat))


# make nice plots using ggplot
colnames(mat) <- c("MES", "ADRN")
matdf     <- as.data.frame(mat)
matdf$SE  <- c("SE in SK-N-AS", "SE in CLB-GA")
matdf     <- gather(matdf, key, val, -SE)
matdf$key <- factor(matdf$key, levels = c("MES", "ADRN"))

m <- ggplot(matdf, aes(key,val, fill = SE)) + geom_col() + 
  scale_fill_manual(values= c(ColNo, ColMes)) +
  ylab("Fraction of interactions") +
  xlab("") + theme(legend.title=element_blank()) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
m
ggsave(filename = params$figure, m, width = 4.0, height = 5)
```


### Legend shortly explained
- *Mesenchymal*: Percentage of interactions ending in enhancer/super enhancer classified as Mesenchymal
- *Non Mesenchymal*: Percentage of interactions ending in enhancer/super enhancer classified as non mesenchymal
- *not classified*: Percentage of interactions ending in an (super) enhancer that belongs to no class but was in the set of consensus enhancers/Super enhancers
- *Not in enhancer*: Percentage of interactions that do not end in any enhancer element but somewhere else


```{R enhancers}
mesMes <- sum((countOverlaps(enhancer[enhancer$Fold > 0], S)))/length(S)
mesNon <- sum((countOverlaps(enhancer[enhancer$Fold > 0], C)))/length(C)
nonMes <- sum((countOverlaps(enhancer[enhancer$Fold < 0], S)))/length(S)
nonNon <- sum((countOverlaps(enhancer[enhancer$Fold < 0], C)))/length(C)


df <- data.frame(mm = mesMes,
           mn  = mesNon,
           nm = nonMes,
           nn = nonNon)

df3 <- data.frame(mm = mesMes*length(S),
           mn  = mesNon*length(C),
           nm = nonMes*length(S),
           nn = nonNon*length(C))


res <- c(mesMes, mesNon, nonMes, nonNon)
res <- res/sum(res)
names(res) <- c("mesE SK-N-AS", "mesE CLB-GA", "non SKNAS", "non CLBGA")
barplot(res, legend =names(res)  )

mSESum <- mesMes + mesNon
mm    <- mesMes/mSESum
mn    <- mesNon/mSESum
nSESum <- nonMes + nonNon
nm <- nonMes/nSESum
nn <- nonNon/nSESum

stacked <-c(mm,mn, nm,nn) 
names(stacked) <- c(rep("mes E",2), rep("non mes E",2))
mat <- matrix(stacked,2)
colnames(mat) <- c("mes E", "non mes E")

barplot(mat, legend = colnames(mat))


# make nice plots using ggplot
colnames(mat) <- c("MES", "ADRN")
matdf         <- as.data.frame(mat)
matdf$SE      <- c("SK-N-AS", "CLB-GA")
matdf         <- gather(matdf, key, val, -SE)
matdf$key     <- factor(matdf$key, levels = c("MES", "ADRN"))

m <- ggplot(matdf, aes(key,val, fill = SE)) + geom_col() + 
  scale_fill_manual(values= c(ColNo, ColMes)) +
  ylab("fraction of interactions") +
  xlab("") + theme(legend.title=element_blank()) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
m

ggsave(filename = params$sup_figure, m, width = 4.0, height = 5)

```
