---
title: "IGV plots"
author: "Health Data Science Unit"
date: "`r date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    number_sections: false
params:
  work_dir     : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/"
  path_config  : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/src/project_NB_SE/scripts/figure3/figure3g_tracks.txt"
  path_high_bed: "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/src/project_NB_SE/scripts/figure3/figure3g_highlight_ranges.bed"
  cells_h      : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/analysis/cells/chipseq/H3K27ac/NMF/cells_consensusSE_K3_Hmatrix_hnorm.RDS"
  tumor_h      : "/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/analysis/tumor/chipseq/H3K27ac/NMF/tumor_consensusSE_K4_Hmatrix_hnorm.RDS"
  width_window : 3000
  ymax  : "max" 
  chr   : "chr11"
  start : 68776977
  end   : 69667477
  name  : "CCND1"
  figure: "results/figure3/figure3g_IGV.pdf"
  width : 8
  height: 4
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
"/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/src/project_NB_SE/scr ipts/figure3/test_tracks.txt"
#this chunk sets global options for Rmarkdown
knitr::opts_chunk$set(echo = TRUE)
#setting working directory. Should be home directory of the project
knitr::opts_knit$set(root.dir=params$work_dir)
options(knitr.table.format="html")


library(tidyverse)
library(viridis)
library(rtracklayer)
library(biovizBase)
library(ggbio)
library(Homo.sapiens)
library(GenomicInteractions)
library(knitr)

NCORES <- 30L
```

## Data loading

```{r helper_functions}
```
  
```{r dataloading}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                             Read data                                      ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
kable(read.table(params$path_config, header = TRUE)) 

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                     assign samples to MES signature                        ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
assign_signature <- function(h){
  is_h <- t(apply(h, 1, function(x){
    k <- kmeans(x, 2)
    max_idx <- which.max(k$centers)
    if_else(k$cluster == max_idx, TRUE, FALSE)
  }))
  colnames(is_h) <- colnames(h)
  rownames(is_h) <- rownames(h)
  is_h
}

cells_h <- readRDS(params$cells_h)
tumor_h <- readRDS(params$tumor_h)

cells_mes <- assign_signature(cells_h)["MES",]
tumor_mes <- assign_signature(tumor_h)["MES",]

cells_mes
tumor_mes
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                     Split tracks between MES and ADRN                      ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
tracksIDs <- c("SK-N-AS HiChIP",
               "SK-N-AS ATAC-seq",
               "SK-N-AS TF footprint",
               "SK-N-AS SE",
               "MES H3K27ac Cell lines",
               "MES H3K27ac tumors",
               "Tumor SE",
               "ADRN H3K27ac tumors",
               "ADRN H3K27ac Cell lines",
               "CLB-GA HiChIP",
               "CLB-GA ATACseq",
               "CLB-GA TF footprint",
               "CLB-GA SE")



tracks_df <- read.table(params$path_config, header = 1, stringsAsFactors = FALSE)
tracks_df



tracks_df$trackID[(tracks_df$trackID == "H3K27ac tumors") & (tracks_df$sampleID %in% names(tumor_mes)[tumor_mes])] <- "MES H3K27ac tumors"
tracks_df$trackID[(tracks_df$trackID == "H3K27ac tumors") & (tracks_df$sampleID %in% names(tumor_mes)[!tumor_mes])] <- "ADRN H3K27ac tumors"
tracks_df$color[tracks_df$trackID == "MES H3K27ac tumors"] <- "#921BBA"
tracks_df$color[tracks_df$trackID == "ADRN H3K27ac tumors"] <- "#B2DF8A"

tracks_df$trackID[(tracks_df$trackID == "H3K27ac Cell lines") & (tracks_df$sampleID %in% names(cells_mes)[cells_mes])] <- "MES H3K27ac Cell lines"
tracks_df$trackID[(tracks_df$trackID == "H3K27ac Cell lines") & (tracks_df$sampleID %in% names(cells_mes)[!cells_mes])] <- "ADRN H3K27ac Cell lines"
tracks_df$color[tracks_df$trackID == "MES H3K27ac Cell lines"] <- "#420A68FF"
tracks_df$color[tracks_df$trackID == "ADRN H3K27ac Cell lines"] <- "#2FB47CFF"


# set sorted levels in the correct tracks order
tracksIDs <- tracksIDs[tracksIDs %in% tracks_df$trackID]
tracks_df$trackID <- factor(tracks_df$trackID, levels = tracksIDs)


# trackIDs <- tracks_df$trackID
# tracks_config <- split(tracks_df, trackIDs)
# names(tracks_config)
```

```{r plot}
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                      Define region of interest                             ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

roi <- GRanges(seqnames = params$chr,
               ranges   = IRanges(start = params$start,
                                  end   = params$end,
                                  name  = params$name))
roi


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                      Reaf regiong to highlight                             ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##


highlight_ranges <- read.table(params$path_high_bed, stringsAsFactors=FALSE)

highlight_ranges <- read.table("/icgc/dkfzlsdf/analysis/B080/crg/B087_Neuroblastoma/publication_GEO/src/project_NB_SE/scripts/figure3/figure3g_highlight_ranges.bed", stringsAsFactors=FALSE)
colnames(highlight_ranges) <- c("chr", "start", "end", "type", "color")
highlight_ranges <- makeGRangesFromDataFrame(highlight_ranges, keep.extra.column = TRUE)
highlight_ranges


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                              Read Track info                               ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
ymax <- params$ymax
ymax <- ifelse(ymax == "max", ymax, as.numeric(ymax))

gg_tracks <- plot_tracks_from_config(tracks_df, width_window = params$width_window)
gg_tracks
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                 Save Plot Tracks                       ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# ggsave(filename = params$figure,
#        plot     = gg_tracks,
#        width    = 8,
#        height   = 4)



```



```{r helper_functions}

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                           basic ggplot theme                               ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
tm <- theme_bw() +
  theme(#axis.line=element_line(colour = "black"),
    #axis.line = element_line(size = .1, colour = "grey20"),
    legend.text = element_text(size=5),
    #axis.line=element_blank(),
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank())


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                  Read BigWig files and make Tracks                         ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
track_BW <- function(file, region2plot, sampleID = "unkown", color, width_window){
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##  Bin region to plot and assign mean score of BigWigs   ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  
  region2plot_bin <- tile(region2plot, width = width_window)
  region2plot_bin <- unlist(region2plot_bin)
  region2plot_bin$sampleID <- rep(sampleID, length(region2plot_bin))
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                       Read file                        ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  bwf <- BigWigFile(file)
  bw <- rtracklayer::import(bwf)
  # keep only region to plot
  bw <- subsetByOverlaps(bw, region2plot)
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##           Find signal in region to plot bins           ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  # center BigWig regions and assign one of the bins
  bw_df <- data.frame(regionCenter = round(start(bw) + width(bw)/2),
                      score = bw$score)
  bw_df$bin <- cut(bw_df$regionCenter, breaks = start(region2plot_bin), labels = FALSE, include.lowest = TRUE)
  bw_df <- bw_df[!is.na(bw_df$bin),]
  # Keep only mean signal per bin
  bw_df <- bw_df %>% 
    group_by(bin) %>% 
    summarise(avg_score = mean(score))
  # Build vector of signal to bind with others BigWigs
  signal <- rep(0, length(region2plot_bin))
  signal[bw_df$bin] <- bw_df$avg_score
  signal[signal <= 0] <- -0.05 
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##          Assign values to binned region                ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  region2plot_bin$score <- signal
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Plot BigWig summary                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ymax <- ifelse(ymax == "max", max(region2plot_bin$score), ymax)
  bw.gg <- ggplot(region2plot_bin, aes(y=score, fill = sampleID)) +
    scale_fill_manual(values = as.character(color)) +
    geom_bar() +
    xlim(region2plot) +
    ylim(c(-1,ymax)) +
    tm #+ 
  #theme(legend.position="none")
  return(bw.gg)
}



##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##             Read BigWig files and make summary Track                       ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

# Aux function
dualBreak <- function(x){
  top <- round(max(x))
  if(top > 10){
    top <- ceiling(top/10) * 10  
  }
  breaks <- c(0,top)
  names(breaks) <- attr(breaks,"labels")
  #message(breaks)
  breaks
}

# files: named character vector with paths
track_summary_BW <- function(files, region2plot, color){
  #message(sampleID) 
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                       Read files                       ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  vals <-mclapply(names(files), function(name){
    file <- files[name]
    # Read bigwig
    bw <- rtracklayer::import(file)
    bw <- subsetByOverlaps(bw, region2plot)
    # center values 
    v <- (values(bw)$score) # as the ranges are the same, ignore these
    m <- round(start(bw) + width(bw)/2)
    val <- data.frame(m = m, v = v, n = rep(name, length(m)))
    
    # split regions into bins
    nBins  <- 800
    breaks <- seq(start(region2plot), end(region2plot), length.out = nBins)
    
    val$bin <- cut(val$m, breaks = breaks, labels = F, include.lowest = T)
    val <- group_by(val, bin) %>% summarise(signal = median(v), mid = breaks[median(bin)])
    val$name <- name
    return(val)
    
  }, mc.cores = NCORES)
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##             Summarize Bigwig signal                    ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ds <- do.call(rbind, vals)
  
  dsD <- ds %>% 
    group_by(mid ) %>% 
    summarise(lowest = quantile(signal)[1],
              lower  = quantile(signal)[2],
              median = median(signal),
              upper  = quantile(signal)[4],
              uppest = quantile(signal)[5])
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##             Smooth with loess                          ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  x <- as.matrix(dsD[,1])
  span <- 0.01
  n    <- 700
  d <- apply(as.matrix(dsD[,2:6]), 2, function(y){
    loess.smooth(x,y, span = span, evaluation = n)$y
  })
  # get x values, using same parameters
  x <-  loess.smooth(x,as.matrix(dsD[,2]), span = span, evaluation = n)$x
  allBWS <- as.tibble(cbind(x,d))
  # allBWS <- dsD
  # allBWS$x <- dsD$mid
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Plot BigWig summary                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  #breaks <- dualBreak(dualBreak(allBWS$upper))
  ymax <- ifelse(ymax == "max", max(allBWS$upper), ymax)
  signal <- ggplot(allBWS, aes(x, upper)) + 
    #geom_segment(data = data.frame(x = median(allBWS$x),ymax =  max(breaks)), aes(x = x, xend  = x, y = 0 , yend = ymax, alpha = 0), color = "white", size= 0) +   
    # geom_ribbon(aes(ymin = lowest, ymax = uppest), fill = "#00afaf" , alpha = 0.5) + 
    geom_ribbon(aes(ymin = lower, ymax = upper), fill = as.character(color) , alpha = 1) + 
    #geom_ribbon(aes(ymin = lowest, ymax = uppest), fill = as.character(color) , alpha = 1) + 
    ylab("") + 
    ylim(c(-1,ymax)) 
    #scale_y_continuous(breaks = c(0,30)) 
    #scale_y_continuous(breaks = breaks) 
  return(signal)
}

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##       Read BigWig files and make summary BigWig Histogram Track            ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

# files: named character vector with paths
track_summary_hist_BW <- function(files, region2plot, color, width_window){
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##  Bin region to plot and assign mean score of BigWigs   ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  region2plot_bin <- tile(region2plot, width = width_window)
  region2plot_bin <- unlist(region2plot_bin)
  #names(region2plot_bin) <- NULL
  
  region2plot_bin$sampleID <- rep("avg_signal", length(region2plot_bin))
  
  # region2plot_bin_df <- as.data.frame(region2plot_bin)
  # head(region2plot_bin_df)
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##             Summarize Bigwig signal                    ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##        Retrive signal over region to plot Bins         ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  
  signal_over_bins <- mclapply(files, function(file){
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    ##                       Read file                        ##
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    bwf <- BigWigFile(file)
    bw <- rtracklayer::import(bwf)
    # keep only region to plot
    bw <- subsetByOverlaps(bw, region2plot)
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    ##           Find signal in region to plot bins           ##
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    # center BigWig regions and assign one of the bins
    bw_df <- data.frame(regionCenter = round(start(bw) + width(bw)/2),
                        score = bw$score)
    bw_df$bin <- cut(bw_df$regionCenter, breaks = start(region2plot_bin), labels = FALSE, include.lowest = TRUE)
    bw_df <- bw_df[!is.na(bw_df$bin),]
    # Keep only mean signal per bin
    bw_df <- bw_df %>% 
      group_by(bin) %>% 
      summarise(avg_score = mean(score))
    # Build vector of signal to bind with others BigWigs
    signal <- rep(0, length(region2plot_bin))
    signal[bw_df$bin] <- bw_df$avg_score
    signal[signal <= 0] <- -0.05 
    return(signal)
  }, mc.cores = NCORES)
  signal_over_bins <- do.call(cbind, signal_over_bins)
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##          Assign values to binned region                ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  region2plot_bin$score <- rowMeans(signal_over_bins)
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Plot BigWig summary                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ymax <- ifelse(ymax == "max", max(region2plot_bin$score), ymax)
  bw.gg <- ggplot(region2plot_bin, aes(y=score, fill = sampleID)) +
    scale_fill_manual(values = as.character(color)) +
    geom_bar() +
    xlim(region2plot) +
    ylim(c(-1,ymax)) +
    tm #+ 
  #theme(legend.position="none")
  return(bw.gg)
}





##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                           Read BED files                                   ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# Read Bed files
track_Bed <- function(file, region2plot, sampleID = "unkown", showlabel.bed = TRUE, color){
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                          Read file                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  bed <- read.table(file, as.is=TRUE)
  if (showlabel.bed) {
    bed <- bed[,1:4]
    colnames(bed) = c('chr','start','end','id')
  } else {
    tmp <-strrep(" ",1:nrow(bed))
    #tmp <- rep(" ", nrow(bed))
    bed <- cbind(bed[,1:3], tmp)
    colnames(bed) = c('chr','start','end','id')
    bed$id = as.character(bed$id)
  }
  
  bed <-  makeGRangesFromDataFrame(bed, keep.extra.column = TRUE)
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Regions of interest                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  bed <- subsetByOverlaps(bed, region2plot)
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                    Plot Bed                            ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  # if plain bed, plot rectangles and exit
  # Placeholder to color
  bed$sampleID <- rep(sampleID, length(bed))
  bed$type <- rep("exon", length(bed))
  SE <- bed
  
  names(bed) <- bed$id
  bed <- split(bed, bed$id)
  bed.gg <- autoplot(bed, aes(type = type),
                     color = as.character(color),
                     fill = as.character(color),
                     #size = .1,
                     label.size = .5,
                     range.geom = "rect") +
    xlim(region2plot) +
    tm +theme(legend.position="none")
  
  if (length(bed) == 0) {
    bed.gg <- ggplot(data.frame()) + geom_point() 
  }
  return(bed.gg)
  
}


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                          Gene model Track                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

getGeneModels <- function(db = Homo.sapiens, regionGR) {
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##         Get genes in the Selected region               ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  # Read organism DB info
  txdb <- OrganismDbi:::.getTxDb(db)
  # Parse all genes that are inside the selected region
  gr <- crunch(txdb, regionGR, 
               truncate.gaps = FALSE,
               columns = c("tx_id", "tx_name", "gene_id"),
               ratio = 0.0025)
  # Find gene SYMBOL
  grl <- split(gr, gr$tx_id)
  lbs <- AnnotationDbi::select(db, names(grl), "SYMBOL", "TXID")
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##               Keep only annotated genes                ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  # Keep only genes with a symbol
  lbs <- lbs[!is.na(lbs$SYMBOL),,drop=FALSE]
  grl <- grl[match(lbs$TXID, names(grl))]
  # keep only the longest transcript
  lbs$len <- unlist(width(GenomicRanges::reduce(grl, ignore.strand=TRUE)))
  lbs <- lbs %>%
    group_by(SYMBOL) %>%
    arrange(desc(len)) %>%
    top_n(1) %>%
    ungroup()
  grl <- grl[match(lbs$TXID, names(grl))]
  names(grl) <- lbs$SYMBOL[match(names(grl), lbs$TXID)] 
  # keep only the longest transcript
  lbs <- data.frame(SYMBOL = names(grl),
                    len = unlist(width(GenomicRanges::reduce(grl, ignore.strand=TRUE)))) %>%
    group_by(SYMBOL) %>%
    mutate(order = order(len, decreasing = TRUE)) %>%
    ungroup()
  grl <- grl[which(lbs$order == 1)]
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                       Plot track                       ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  gg_roi <- autoplot(grl, aes(type = type),
                     color = inferno(5)[4],
                     fill = inferno(5)[4],
                     #size = .1,
                     label.size = .5,
                     exon.rect.h = 0.05,
                     cds.rect.h  = 0.05,
                     utr.rect.h  = 0.05/2,
                     group.selfish = FALSE,
                     #stat = "stepping", 
                     #arrow.rate = 0.015,
                     #length = unit(0.1, "points"),
                     range.geom = "rect",
                     rect.height = 0.05/4,
                     gap.geom = "arrow") +
    tm + theme(legend.position="none")
}


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                           Read BEDPE files                                 ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# Read Bed files
track_Bedpe <- function(file, region2plot, sampleID = "unkown", color, FDR = 0.25){
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                          Read file                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  hic_data <- makeGenomicInteractionsFromFile(file, 
                                              type="bedpe", 
                                              experiment_name = name, 
                                              description = name)
  meta <- read.table(file, colClasses = c(rep("NULL", 7), "numeric"))
  colnames(meta) <- "FDR"
  hic_data$FDR <- meta$FDR
  bedpe <- hic_data[hic_data$FDR <= FDR]
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Regions of interest                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  #print(bedpe)
  bedpe <- subsetByOverlaps(bedpe, region2plot)
  bedpe <- subsetByOverlaps(bedpe, highlight_ranges)
  print(bedpe)
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Plot BEDPE.                             ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    hichip <- anchorOne(bedpe)
  start(hichip) <- start(hichip) + floor(width(hichip)/2)
  end(hichip)   <- start(anchorTwo(bedpe)) + floor(width(anchorTwo(bedpe))/2)
  
  
  hichip$FDR  <- -log10(bedpe$FDR)
  hichip$rank <-  if (length(bedpe) > 0) rank(hichip$FDR)+3 else numeric()
  #hichip$rank <-  if (length(bedpe) > 0) order(bedpe$FDR)+3 else numeric()
  
  hichip <- hichip[start(hichip) >= start(region2plot)]
  hichip <- hichip[end(hichip) <= end(region2plot)]
  
  #print(as.data.frame(hichip))
  #print(summary(hichip$FDR))
  #print(summary(-log10(hichip$FDR)))
  print(hichip)
  #print(class(hichip))
  
  
  bedpe.gg <- ggplot(hichip) +
    #geom_arch(hichip, aes(color=FDR, height=rank), size=.4) + 
    geom_arch(hichip, aes(height=rank), color=color, size=.4) + 
    #scale_size_continuous(limits = c(0,1)) +
    #scale_color_gradientn(colours = inferno(100, end = 0.9), limits=c(-log10(0.25),10)) +
    #scale_y_continuous(limits = c(0,1)) +
    xlim(region2plot) +
    tm +
    theme(axis.text.y  = element_blank(),
          axis.ticks.y = element_blank()) 
  #readBedpe(path, name = sampleID, col = color, FDR = 0.25)
  
  #return(NULL)
  
}

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                   Read footprint and print Track                           ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
# Read Bed files
track_footprint <- function(file, highlight_ranges, region2plot, sampleID = "unkown", showlabel.bed = TRUE, tf_activity = NA){
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                          Read file                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  bed <- read.table(file, as.is=TRUE)
  if (showlabel.bed) {
    bed <- bed[,1:4]
    colnames(bed) = c('chr','start','end','id')
  } else {
    tmp <-strrep(" ",1:nrow(bed))
    bed <- cbind(bed[,1:3], tmp)
    colnames(bed) = c('chr','start','end','id')
    bed$id = as.character(bed$id)
  }
  bed <- bed[!is.na(bed$chr),,drop=FALSE]
  bed <-  makeGRangesFromDataFrame(bed, keep.extra.column = TRUE)
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Regions of interest                     ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  bed <- subsetByOverlaps(bed, highlight_ranges)
  bed <- unlist(GenomicRanges::reduce(split(bed + 1000, bed$id )))
  bed$id <- names(bed)
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                Find TF MES activity                    ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  if (is.na(tf_activity)) {
    bed$tf_activity <- 1:length(bed)
  } else {
    # Drop TFs without activity score
    bed <- bed[bed$id %in% rownames(tf_activity)]
    # Summarize TF activity
    tf.mes <- tf_activity$Activity.Mes[match(bed$id, rownames(tf_activity))] 
    bed$tf_activity <- tf.mes
    bed$tf_Mes      <- tf.mes > 1
  }
  
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  ##                    Plot footprint                      ##
  ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
  #return(bed)
  bed.gg <- ggplot(bed) +
    stat_identity(aes(x = start, y = tf_activity, color = tf_activity), geom = "point") +
    geom_text(aes(x = start, y = tf_activity, label=id), size=1, color="black") +
    scale_color_gradient2(low = viridis(5)[4], mid = "white",
                           high = inferno(5)[2], midpoint = 0, space = "Lab",
                           na.value = "grey50", guide = "colourbar", aesthetics = "color") +
    xlim(region2plot) +
    scale_y_continuous(breaks= scales::pretty_breaks(n=3)) +
    tm +
    theme(legend.position="none")
  
  
  return(bed.gg)
  
  if (length(bed) == 0) {
    bed.gg <- ggplot(data.frame()) + geom_point() 
  }
  return(bed.gg)
  
}


##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                           Highlight Track                                  ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
track_highlight <- function(highlight_ranges, region2plot){
  highlight_ranges <- subsetByOverlaps(highlight_ranges, region2plot)
  ggplot(highlight_ranges, aes(fill = type)) + 
    geom_rect(color = NA, alpha = 0.3) +
    scale_fill_manual(values = c("grey80", "#921BBA", "#B2DF8A")) +
    xlim(region2plot) +
    tm


}

##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##                             Plot Tracks                                    ##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
##––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––##

plot_tracks_from_config <- function(samples, width_window){
  # Read file with paths and ID of tracks to plot
  #samples <- read.table(path_config, header = TRUE, stringsAsFactors = FALSE)
  #print(samples)
  #x <- t(samples)
  
  #the tracks should be in the correct order in the config file
  trackIDs <- samples$trackID
  tracks_config <- split(samples, trackIDs)
  
  lapply(names(roi), function(region2plotID){
    region2plot <- roi[region2plotID]
    message(names(region2plot))
    
    track2plot.gg <- lapply(tracks_config, function(track_config){
      #order  cond 
      type     <- unique(track_config$type)
      sampleID <- track_config$sampleID
      path     <- setNames(track_config$path, track_config$sampleID)
      labels   <- as.logical(track_config$labels)
      color    <- unique(track_config$color)
      
      print(paste(paste(sampleID, collapse = " "),
                  "",
                  type))
      # print(labels)
      # print(samp["labels"])
      # 
      # print(as.character(sampleID))
      # print(type)
      
      # start <- start(region2plot) - 0
      # end   <- end(region2plot)   + 0
      
      if (type == "bw") {
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        ##                Plot BigWigs                            ##
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        bw.gg <- track_BW(file = path, region2plot, sampleID = sampleID, color, width_window)
        return(bw.gg)
        
      } else if (type == "bed.fp" | type == "bed") {
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        ##                Plot Beds                               ##
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        bed.gg <- track_Bed(file = path, region2plot, sampleID = sampleID, showlabel.bed = labels, color)
        return(bed.gg)
        
      } else if (type == "sumbw") {
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        ##                Plot BigWig summary                     ##
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        gg_bws <- track_summary_BW(files = path, region2plot, color)
        #gg_bws <- track_summary_hist_BW(files = path, region2plot, color, width_window)
        return(gg_bws)
        
      } else if (type == "bedpe") {
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        ##                Plot BigWig summary                     ##
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        gg_bp <- track_Bedpe(file = path, region2plot, sampleID = sampleID, color, FDR = 0.25)
        return(gg_bp)
        
      } else if (type == "footprint") {
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        ##                Plot BigWig summary                     ##
        ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
        gg_fp <- track_footprint(file = path, highlight_ranges, region2plot, sampleID = sampleID,
                        showlabel.bed = TRUE, tf_activity = NA)
        return(gg_fp)
        
      } else {
        return(NULL)
      }
    })
    #}, mc.cores = NCORES)
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    ##                Add highlight track                     ##
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    track2plot.gg <- c(Highlight = track_highlight(highlight_ranges, region2plot),
                       track2plot.gg)
    
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    ##                Add gene track                          ##
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    track2plot.gg$`Gene Region` <- getGeneModels(Homo.sapiens, region2plot)
    
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    ##                Plot Tracks                             ##
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    # Heights of each track
    track2plot.gg$heights <- c(1, 
                               sapply(tracks_config, function(x) unique(x$height)) , 
                               2)
    # Parameters to change track label colors
    track2plot.gg$label.bg.fill    <- "Grey90"
    track2plot.gg$label.text.color <- "Grey20"
    track2plot.gg$label.text.cex   <- 0.5
    track2plot.gg$label.text.angle <- 0
    track2plot.gg$label.width      <- unit(5.0, "lines")
    track2plot.gg$xlim <- region2plot
    track2plot.gg <- track2plot.gg[!sapply(track2plot.gg, is.null)] 
    print("one step")
    
    # make tracks object to plot
    tracks.r <- do.call(tracks, track2plot.gg)  + 
      tm + 
      theme(legend.position="none",
            axis.text.y = element_text(size=5))
    
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    ##                 Save Plot Tracks                       ##
    ##––––––––––––––––––––––––––––––––––––––––––––––––––––––––##
    ggsave(filename = params$figure,
       plot     = tracks.r,
       width    = params$width,
       height   = params$height)

    return(tracks.r)
    
  })
  
}




```


